ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,668 @@
+
+Updating to this new version from `S1` with reprocessing starting at `L10`,
+results in a workflow instance with key `3` in the state. Then, resprocessing
+`L11` is skipped again (there is still no procesor for this record). And lastly,
+when processing `L12`, there is a failure, because the workflow instance with
+key `2` cannot be found.
+
+**In the engine proposed by this document**, the state of `S1` is copied and
+replay starts at `L10`. We need to note additionally, that `L10` and `L12` are
+commands, while `L11` is an event.
+
+First, `L10` is skipped, because commands are skipped during replay. Then,
+during the replay of `L11`, a workflow instance is stored in the state with key
+`2` as described by the data inside `L11`. Finally, processing starts again at
+`L12`, during which the Process BPMN element of workflow `1` in workflow
+instance `2` transitions to `Element_Activated`.
+
+Now, let's again make the same imaginary bug fix in the key generation. From
+now, only odd numbers are generated.
+
+Updating to this new version from `S1` with replay starting at `L10` results is
+the same state as before the update. First `L10` is skipped, because commands
+are skipped during replay. Then, during the replay of `L11`, a workflow instance
+is stored in the state with key `2` as described by the data inside `L11`. Note,
+that this key is not newly generated, in which case it would've had the value
+`3`. Finally, processing starts again at `L12`, during which the Process BPMN
+element of workflow `1` in workflow instance `2` transitions to
+`Element_Activated`.
+
+Note, that in this case the workflow instance was correctly referred to from
+`L12`, because the events were directly applied as changes to the state instead of reapplying the business logic (which had changed).
+
+This example also shows that the effect of this proposal is that bug fixes in
+the form of business logic changes can no longer be applied to the processing of
+records that have already been processed previously. Their situation will not be
+altered since their records have already been written to the log stream in the
+form of events.
+ 
+## Key Generation
+
+The failure in the previous example occurred in key generation. Even though this
+proposal improves dealing with changes to keys, we should consider keys in more
+detail.
+
+A key (i.e. a long value) identifies an entity (e.g. a workflow instance) uniquely across all partitions. Most records have a key, except for user commands that don't refer to an existing resource (e.g. create workflow instance command). When a record is written for a new entity when processing a command, then a new key is generated.
+
+On replay, the key generator must be restored to generate new keys on processing. To restore the key generator, we read all records on the stream to find the latest key (i.e. the highest long value). Note, that the last record on the stream might not hold the latest key, because it can refer a previous entity.
```

Comment:
What do you mean by key namespace? AFAIK, only 1 key generator is used to produce all keys within a partition, but all keys unique because they are prefixed by the unique partition key.

Reply:
Ok, so we have nested key namespaces. The outer layer are all keys prefixed by the partition key. The inner layer is the keys per partition which are across all entities.